crates/evm/evm/src/executors/invariant/error.rs

```diff
@@ -202,70 +224,137 @@ impl InvariantFuzzError {
-    /// same process again.
-    ///
-    /// Returns the smallest sequence found.
+    /// If the number of calls is small enough, we can guarantee maximally shrink
     fn try_shrinking<'a>(
         &self,
         calls: &'a [BasicTxDetails],
         executor: &Executor,
     ) -> Vec<&'a BasicTxDetails> {
-        let mut anchor = 0;
-        let mut removed_calls = vec![];
-        let mut shrunk = calls.iter().collect::<Vec<_>>();
         trace!(target: "forge::test", "Shrinking.");
+        let seq_idxs = self.try_shrinking_recurse(calls, executor, 0, 0);
 
-        while anchor != calls.len() {
-            // Get the latest removed element, so we know which one to remove next.
-            let removed =
-                match self.fails_successfully(executor.clone(), calls, anchor, &removed_calls) {
-                    Ok(new_sequence) => {
-                        if shrunk.len() > new_sequence.len() {
-                            shrunk = new_sequence;
-                        }
-                        removed_calls.last().cloned()
-                    }
-                    Err(_) => removed_calls.pop(),
-                };
-
-            if let Some(last_removed) = removed {
-                // If we haven't reached the end of the sequence, then remove the next element.
-                // Otherwise, restart the process with an incremented anchor.
-
-                let next_removed = last_removed + 1;
+        calls
+            .iter()
+            .enumerate()
+            .filter_map(|(i, call)| if seq_idxs.contains(&i) { Some(call) } else { None })
+            .collect()
+    }
 
-                if next_removed > calls.len() - 1 {
-                    anchor += 1;
-                    removed_calls = vec![];
-                    continue
-                }
+    /// We try to construct a [powerset](https://en.wikipedia.org/wiki/Power_set) of the sequence if
+    /// the configuration allows for it and the length of the sequence is small enough. If we
+    /// do construct the powerset, we run all of the sequences in parallel, finding the smallest
+    /// one. If we have ran the powerset, we are guaranteed to find the smallest sequence for a
+    /// given set of calls (however, not guaranteed *global* smallest if a random sequence was
+    /// used at any point during recursion).
+    ///
+    /// If we were unable to construct a powerset, we construct a random sample over the sequence
+    /// and run these sequences in parallel instead.
+    ///
+    /// After running either the powerset or the random sequences, we check if we successfully
+    /// shrunk the call sequence.
+    fn try_shrinking_recurse(
+        &self,
+        calls: &[BasicTxDetails],
+        executor: &Executor,
+        runs: usize,
+        retries: usize,
+    ) -> Vec<usize> {
+        let shrunk = Arc::new(RwLock::new((0..calls.len()).collect()));
```

Comment:
what does this represent? can we give it a more descriptive name?

Reply:
indices that should be included (use_call_indices). will fix